core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala

```diff
@@ -475,7 +475,148 @@ class DAGSchedulerSuite
+    val reduceRdd = new MyRDD(sc, 2, List(shuffleDep))
+    submit(reduceRdd, Array(0, 1))
+    
+    complete(taskSets(0), Seq(
+      (Success, makeMapStatus("hostA", 1)),
+      (Success, makeMapStatus("hostB", 1))))
+    
+    for (x <- 1 to Stage.MAX_STAGE_FAILURES) {
+      // the 2nd ResultTask failed
+      complete(taskSets(1), Seq(
+        (Success, 42),
+        (FetchFailed(makeBlockManagerId("hostA"), shuffleId, 0, 0, "ignored"), null)))
+
+      scheduler.resubmitFailedStages()
+      if (x < Stage.MAX_STAGE_FAILURES) {
+        assert(scheduler.runningStages.nonEmpty)
+        assert(!ended)
+      } else {
+        // Stage has been aborted and removed from running stages
+        assertDataStructuresEmpty()
+        sc.listenerBus.waitUntilEmpty(1000)
+        assert(ended)
+        assert(jobResult.isInstanceOf[JobFailed])
+      }
+    }
+  }
+
+
+  test("Multiple consecutive Fetch failures in a stage triggers an abort.") {
+    // Create a new Listener to confirm that the listenerBus sees the JobEnd message 
+    // when we abort the stage. This message will also be consumed by the EventLoggingListener
+    // so this will propagate up to the user. 
+    var ended = false
+    var jobResult : JobResult = null
+    class EndListener extends SparkListener {
+      override def onJobEnd(jobEnd: SparkListenerJobEnd): Unit = {
+        jobResult = jobEnd.jobResult
+        ended = true
+      }
+    }
+
+    sc.listenerBus.addListener(new EndListener())
+
+    val shuffleMapRdd = new MyRDD(sc, 8, Nil)
+    val shuffleDep = new ShuffleDependency(shuffleMapRdd, null)
+    val shuffleId = shuffleDep.shuffleId
+    val reduceRdd = new MyRDD(sc, 8, List(shuffleDep))
+    submit(reduceRdd, Array(0, 1, 2, 3, 4, 5, 6, 7))
+    
+    complete(taskSets(0), Seq(
+      (Success, makeMapStatus("hostA", 1)),
+      (Success, makeMapStatus("hostA", 1)),
+      (Success, makeMapStatus("hostA", 1)),
+      (Success, makeMapStatus("hostA", 1)),
+      (Success, makeMapStatus("hostA", 1)),
+      (Success, makeMapStatus("hostA", 1)),
+      (Success, makeMapStatus("hostA", 1)),
+      (Success, makeMapStatus("hostB", 1))))
+
+    complete(taskSets(1), Seq(
+      (Success, 42),
+      (FetchFailed(makeBlockManagerId("hostA"), shuffleId, 0, 0, "ignored"), null),
+      (FetchFailed(makeBlockManagerId("hostA"), shuffleId, 0, 0, "ignored1"), null),
+      (FetchFailed(makeBlockManagerId("hostA"), shuffleId, 0, 0, "ignored2"), null),
+      (FetchFailed(makeBlockManagerId("hostA"), shuffleId, 0, 0, "ignored3"), null),
+      (FetchFailed(makeBlockManagerId("hostA"), shuffleId, 0, 0, "ignored4"), null),
+      (FetchFailed(makeBlockManagerId("hostA"), shuffleId, 0, 0, "ignored5"), null),
+      (FetchFailed(makeBlockManagerId("hostA"), shuffleId, 0, 0, "ignored6"), null)))
+
+    scheduler.resubmitFailedStages()
+    assertDataStructuresEmpty()
+    sc.listenerBus.waitUntilEmpty(1000)
+    assert(ended)
+    assert(jobResult.isInstanceOf[JobFailed])
```

Comment:
I think in this case, we actually want to make sure the job *succeeds*.  The idea is that after a bunch of fetch failures on attempt 0, we shouldn't give up.  Attempt 1 should still get tried.  And if our tests has all tasks in attempt 1 complete successfully, the job should pass.